File: VOID0000-api/server/vmd/metrics.js
const counters = {
  persistent_cache_hits: 0,
  persistent_cache_misses: 0,
  persistent_cache_corrupt: 0,
  persistent_cache_read_failures: 0,
  persistent_cache_write_failures: 0,
  transforms_generated: 0,
  queue_full: 0,
  queue_timeouts: 0,
};

export function incrementVmdMetric(name) {
  if (!Object.hasOwn(counters, name)) {
    throw new TypeError(`Unknown VMD metric: ${name}`);
  }
  counters[name] += 1;
}

export function getVmdMetricsSnapshot() {
  return { ...counters };
}

File: VOID0000-api/vmd/metrics.go
package vmd

import "sync/atomic"

type Metrics struct {
	persistentCacheHits         atomic.Uint64
	persistentCacheMisses       atomic.Uint64
	persistentCacheCorrupt      atomic.Uint64
	persistentCacheReadFailures atomic.Uint64
	persistentCacheWriteFailure atomic.Uint64
	transformsGenerated         atomic.Uint64
	queueFull                   atomic.Uint64
	queueTimeouts               atomic.Uint64
}

func (m *Metrics) Snapshot(queue QueueSnapshot) map[string]any {
	return map[string]any{
		"persistent_cache_hits":           m.persistentCacheHits.Load(),
		"persistent_cache_misses":         m.persistentCacheMisses.Load(),
		"persistent_cache_corrupt":        m.persistentCacheCorrupt.Load(),
		"persistent_cache_read_failures":  m.persistentCacheReadFailures.Load(),
		"persistent_cache_write_failures": m.persistentCacheWriteFailure.Load(),
		"transforms_generated":            m.transformsGenerated.Load(),
		"queue_full":                      m.queueFull.Load(),
		"queue_timeouts":                  m.queueTimeouts.Load(),
		"work_queue":                      queue,
	}
}
